assets/js/modules/analytics-4/datastore/properties.js

```diff
@@ -499,13 +497,16 @@ const baseActions = {
 		const { googleTagAccountID, googleTagContainerID, googleTagID } =
 			response;
 
-		registry
-			.dispatch( MODULES_ANALYTICS_4 )
-			.setGoogleTagAccountID( googleTagAccountID );
-		registry
-			.dispatch( MODULES_ANALYTICS_4 )
-			.setGoogleTagContainerID( googleTagContainerID );
-		registry.dispatch( MODULES_ANALYTICS_4 ).setGoogleTagID( googleTagID );
+		// Note that when plain actions are dispatched in a generator function action, they are handled asynchronously when they would normally be synchronous.
+		// This means that following the usual pattern of dispatching individual setter actions for the `googleTagAccountID`, `googleTagContainerID` and `googleTagID` settings
+		// each resulted in a rerender of the GoogleTagIDMismatchNotification component, thus resulting in an erroneous call to the GET:container-destinations endpoint with
+		// mismatched settings. To mitigate this, we dispatch a single action here to set all these settings at once. The same applies to the `setSettings()` call above.
```

Comment:
We may want to clarify this part for accuracy. I think what's happening is simply that each set-action is triggering the datastore to notify subscribers which will happen every time state changes. I think the batched update is fine here but I don't think it is because of the way the actions are being called (from a generator or not). Also, these actions are being dispatched from the registry rather than yielded which is handled differently, although I don't think it would make a difference in this circumstance.

Reply:
Hey @aaemnnosttv, thanks for calling this one out. It's made me go and dig into it a bit further with interesting and somewhat surprising results (although, it's still about as odd as it already was).

Essentially, when looking into this the first time, I noticed that when running those same individual setter actions from a regular click handler, they were behaving synchronously, i.e. the component was not being re-rendered between each dispatch. This was not the case within the generator, leading me to my initial conclusion.

Having taken a closer look, what I've found is that the actions are behaving synchronously when dispatched (or yielded) from a function or generator _which has not awaited_. It's a bit strange. Adding an `await` to either a regular async function or via yielding to an async action, changes the behaviour and these same actions are then handled asynchronously with a render between each dispatch.

In order to test and illustrate the point, I've created a [draft PR](https://github.com/google/site-kit-wp/pull/6814) with a test widget and `@wordpress/data` constructs, and this includes a couple of screenshots of some console output in these scenarios.

I've not managed to get to the very root of this behaviour, and am hesitant to go down too much of a rabbit hole here, digging into the Redux middlewares, the event loop, job scheduling and so forth.

So, what I've done is update this comment to reflect the observed facts on the ground. Please take a look, I'm interested to know if you've got any further thoughts on this one.